[NONEVM-916] logpoller log processing & decoding

reductionista opened the pull request:
JIRA: [LINK-916](https://smartcontract-it.atlassian.net/browse/NONEVM-916)

This adds the implementation of `lp.Process()`, and also initializes LogPoller with a client that gets passed along to initialize the EncodedLogCollector.

In order, the main 3 tasks performed by `lp.Process` are:
1. Filter out any logs not matching registered filters
2. Decode each matching log, and extract the subkeys
3. Save matching logs along with their subkey values to the db

reductionista committed:
Add client & collector to LogPoller struct

Also:
- Add GetBlocksWithOpts to Reader
- Update loader.RPClient to be consistent with client.Reader interface

reductionista committed:
Re-generate RPClient & ReaderWriter mocks

reductionista committed:
Update loader tests for interface

reductionista committed:
Add ILogpoller interface and call NewLogPoller() from NewChain()

Note: this will hopefully be renamed to LogPoller later, once we
find a better name for the struct. (logPoller, for consistency with evm?)
and the PR's this depends on have been merged, to avoid merge conflicts

reductionista committed:
fix orm_test.go

reductionista committed:
Fix lints in log_poller.go & types.go

reductionista committed:
Fix lints in log_poller.go

reductionista committed:
WIP

reductionista committed:
Add BlockTime & Program to Event data passed to Process()

reductionista committed:
Use EventTypeProvider interface to represent a codec function for retrieving event types from idl

reductionista committed:
remove unimplemented test

reductionista committed:
Regenerate mock_orm.go

reductionista committed:
Fix lints

reductionista committed:
Add IndexedValue type for converting indexable types to be stored in db

reductionista committed:
Fill in log.ExpiresAt

reductionista committed:
Add sequence # tracking, fix floating point encoding, add tests for NewIndexedValue

reductionista committed:
Fix tests

reductionista committed:
Remove mock_orm.go

reductionista committed:
Use UTC time, and fix lints

reductionista committed:
Add tests - WIP

reductionista committed:
Refactor to use new codec api, based on review

dhaidashenko commented on pkg/solana/chain.go:
```diff
@@ -304,10 +308,14 @@ func newChain(id string, cfg *config.TOMLConfig, ks core.Keystore, lggr logger.L
-		tc = internal.NewLoader[client.ReaderWriter](func() (client.ReaderWriter, error) { return ch.multiNode.SelectRPC() })
-		bc = internal.NewLoader[monitor.BalanceClient](func() (monitor.BalanceClient, error) { return ch.multiNode.SelectRPC() })
+		// TODO: Can we just remove these? They nullify the lazy loaders initialized earlier, don't they?
+		//lc = internal.NewLoader[client.Reader](func() (client.Reader, error) { return ch.multiNode.SelectRPC() })
```
@DylanTinianov do you recall reason to add this?

DylanTinianov replied:
We must still use these internal loaders, since the above lazy loaders will not properly use MultiNode functionality as they hold onto a single RPC rather than using MultiNode selection on each `getClient()` call.

Looking at this now, we can probably remove the following from ch.getClient() since MultiNode should just use the internal loaders and not lazy loaders:

`	if c.cfg.MultiNode.Enabled() {
		return c.multiNode.SelectRPC()
	}
`

reductionista replied:
Ah, I see... the lazy loader would only call `multiNode.SelectRPC()` once, then get stuck on whichever rpc client is returned? Makes sense.

dhaidashenko replied:
Duplicating here my comment from the MultiClient PR. 
We can't remove the `if c.cfg.MultiNode.Enabled() { return c.multiNode.SelectRPC() }` logic from chain's getClient method as getClient is used in other methods of the chain type.

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -100,6 +113,15 @@ func (fl *filters) RegisterFilter(ctx context.Context, filter Filter) error {
+		Events: []codec.IdlEvent{filter.EventIdl.IdlEvent},
+		Types:  filter.EventIdl.IdlTypeDefSlice,
+	}
+	filter.decoder, err = codec.NewIDLEventCodec(idl, binary.LittleEndian())
```
we should do it before Insert to ensure that db always has valid filters

reductionista replied:
I changed this, but ended up having to change it back due to other refactoring.  It's stored in `fl.decoders[ID]` now, but needs to get the ID from inserting the db row first.

Unfortunately, having the decoder as part of the `filter` object broke all of the tests which depend on `assert.Equal(filter, expectedFilter)`. I tried fixing that in various ways, but in the end decided it was cleaner to just have the decoders in a separate index instead of a part of the filter.

The order shouldn't matter, since the filters mutex is held throughout this process.

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -118,6 +140,21 @@ func (fl *filters) RegisterFilter(ctx context.Context, filter Filter) error {
+		fl.knownPrograms[programID]++
+	}
+
+	discriminatorHead := filter.Discriminator()[:10]
```
Are we sure that Distriminator has at least 10 bytes? Should we add validation?

reductionista replied:
The `utils.Discriminator` explicitly returns [8]byte, and the `filter.Discriminator` wrapper around that just calls `base64.EncodeToString` on that which will always return []byte backed by [12]byte array. 

I can make this a bit more explicit by changing the return type of `filter.Discriminator` to `[12]byte`, since we have to generate a new slice of it anyway (at least in this case).

reductionista replied:
I forgot that the return type is `string` not `[]byte`. So instead of changing the return type, I added an AssumptionViolation check that panics if the base64 encoder doesn't encode the 8 bytes passed to it as a 12-byte base64 string. I think the panic makes sense here, because the only way this could potentially happen is during development, if we're updating to a new version of `go` and they've broken backwards compatibility in `encoder/base64` or introduced a bug.  The panic will let us know immediately that we need to roll back or update our code accordingly.

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -118,6 +140,21 @@ func (fl *filters) RegisterFilter(ctx context.Context, filter Filter) error {
 	}
+
+	programID := filter.Address.ToSolana().String()
+	if _, ok = fl.knownPrograms[programID]; !ok {
```
No need for if else statement we can rely on default values and just increment it

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -180,6 +217,26 @@ func (fl *filters) removeFilterFromIndexes(filter Filter) {
+		}
+	}
+
+	discriminatorHead := filter.Discriminator()[:10]
```
Same here, we should add validation to avoid panic

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -210,6 +267,38 @@ func (fl *filters) MatchingFilters(addr PublicKey, eventSignature EventSignature
+	// discriminators if the first 10 characters don't match. If it passes that initial test, we base64-decode the
+	// first 11 characters, and use the first 8 bytes of that as the event sig to call MatchingFilters. The address
+	// also needs to be base58-decoded to pass to MatchingFilters
+	if _, ok := fl.knownDiscriminators[event.Data[:10]]; !ok {
```
Need validation that data is large enough.

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -210,6 +267,38 @@ func (fl *filters) MatchingFilters(addr PublicKey, eventSignature EventSignature
+	}
+	eventSig := EventSignature(decoded[:8])
+
+	return fl.MatchingFilters(PublicKey(addr), eventSig)
```
There is no need for MatchingFilters to be public

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -32,6 +40,11 @@ func newFilters(lggr logger.SugaredLogger, orm ORM) *filters {
 }
 
+func (fl *filters) IncrementSeqNums(filterID int64) int64 {
+	fl.seqNums[filterID]++
```
It might be nice to ensure that filters were loaded

reductionista replied:
I don't think this is the right place for that check, because its only purpose is for `lp.Process` to be able to increment the sequence number each time a new log is processed.

I should probably add a check to `lp.Process`, since we shouldn't be processing logs yet if filters are not loaded. The main loop of `Process` is iterating through the sequence returned by `matchingFilters()`, and I notice this already logs a critical error if it's called before filters are loaded:
```
Invariant violation: expected filters to be loaded before call to matchingFilters
```
But checking prior to this call is better, since then it can return an error.

Looking over this code reminded me, I still need to add a call to `lp.collector.Start()` just after `lp.loadFilters()` returns successfully

reductionista replied:
Oh right, nevermind I forgot there is no seq_num field in filter table. No need for the second one :-)

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -180,6 +217,26 @@ func (fl *filters) removeFilterFromIndexes(filter Filter) {
 		delete(fl.filtersByAddress, filter.Address)
 	}
+
+	programID := filter.Address.ToSolana().String()
```
we should also delete sequence counter

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+			ChainID:        lp.orm.ChainID(),
+			LogIndex:       makeLogIndex(blockData.TransactionIndex, blockData.TransactionLogIndex),
+			BlockHash:      Hash(blockData.BlockHash),
+			BlockNumber:    int64(blockData.BlockHeight),
```
Should we use the slot number instead? BlockHeight defines how many nonempty slots are behind this block. And it's not used by RPC API or explorers, so does not provide any debugging value

reductionista replied:
Oh, I thought block number and block height were the same (counting only actual blocks) while slot number counts the empty slots that don't include any block as well.  Is block number just another name for slot number then?

dhaidashenko replied:
Your understanding is correct. But BlockHeight is not useful for debugging as we can't use it in explorer or in RPC API. All the calls use SlotNumber.

dhaidashenko commented on pkg/solana/logpoller/job.go:
```diff
@@ -114,12 +115,18 @@ func (j *getTransactionsFromBlockJob) Run(ctx context.Context) error {
 
 	if len(block.Transactions) != len(blockSigsOnly.Signatures) {
-		return fmt.Errorf("block %d has %d transactions but %d signatures", j.slotNumber, len(block.Transactions), len(blockSigsOnly.Signatures))
+		return fmt.Errorf("block %d has %d transactions but %d signatures", block.BlockHeight, len(block.Transactions), len(blockSigsOnly.Signatures))
```
We should use slot number. Block Height does not allow us to lookup block in the explorer or fetch it via API.

EasterTheBunny replied:
+1

dhaidashenko commented on pkg/solana/logpoller/loader.go:
```diff
@@ -170,7 +170,7 @@ func (c *EncodedLogCollector) runSlotPolling(ctx context.Context) {
 
 			// not to be run as a job, but as a blocking call
-			result, err := c.client.GetLatestBlockhash(ctxB, rpc.CommitmentFinalized)
+			result, err := c.client.LatestBlockhash(ctxB)
```
LatestBlockhash return block hash with a configured commitment. Default value is `CommitmentConfirmed` it might be better to control it on LP level as TXM or other components might have different requirements

reductionista replied:
Hmm... does the txm & logpoller have to share the same multinode client? Or could it make sense to have one for each? I'd prefer to just configure it once for finalized up front rather than having to keep setting finalized every time. Forgetting to set it in one place could easily cause an issue we don't notice for a while.

reductionista replied:
Also, I would think that using unfinalized blocks would lead to similar issues with the tx mgr. But I guess I should check with the ChainWriter team on that.

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -2,40 +2,64 @@ package logpoller
+	SelectSeqNums(ctx context.Context) (map[int64]int64, error)
+}
+
+type ILogPoller interface {
```
Why do we need to define interface?

reductionista replied:
I think we'll at least need it so we can generate a mock LogPoller that chainreader can use for their unit tests. But it also helps clearly define what the LogPoller API is. Eventually, once we standardize the API across chains, this should get moved into chainlink-common but for now the EVM and Solana one don't line up perfectly.

That reminds me, I wanted to name it `LogPoller` and rename the struct to `logPoller` so it's more similar to what we have in EVM. But at the time, I was working on top of several branches of open PR's, and didn't want to introduce any additional merge conflicts by renaming the struct everywhere.

I'm open to taking it out for now, but I have a feeling we'll have to put it back in at some point, either for the mock or other reasons. What do you think?

EasterTheBunny replied:
Creating unused mocks just adds unnecessary code and brings our code coverage down with no value added. Ideally, mocks should be small and isolated to ONLY what a component needs as a dependency. Since the `LogPoller` doesn't use this interface as a dependency, it shouldn't be here.

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+	if err != nil {
+		return err
+	}
+	lp.collector = NewEncodedLogCollector(cl, lp, lp.lggr)
```
Seems like we need to implement proper client that wrapps multinode or use lazy-loader instead. Passing raw client directly to the loader drop MultiNode functionality.

dhaidashenko replied:
I've created a wrapped client in [this PR](https://github.com/smartcontractkit/chainlink-solana/pull/1006), so that we can seamlessly interact with multinode client or legacy lazy-loaded client.

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
 }
 
+func makeLogIndex(txIndex int, txLogIndex uint) int64 {
+	if txIndex < 0 || txIndex > math.MaxUint32 || txLogIndex > math.MaxUint32 {
```
nit: if txIndex is equal to MaxUint32, we'll have an overflow. We should check that `txIndex > math.MaxInt32`

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
 
+func makeLogIndex(txIndex int, txLogIndex uint) int64 {
+	if txIndex < 0 || txIndex > math.MaxUint32 || txLogIndex > math.MaxUint32 {
+		panic(fmt.Sprintf("txIndex or txLogIndex out of range: txIndex=%d, txLogIndex=%d", txIndex, txLogIndex))
```
Does it make sense to kill the whole CL Node in case of the issues in Solana?

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+	if txIndex < 0 || txIndex > math.MaxUint32 || txLogIndex > math.MaxUint32 {
+		panic(fmt.Sprintf("txIndex or txLogIndex out of range: txIndex=%d, txLogIndex=%d", txIndex, txLogIndex))
+	}
+	return int64(math.MaxUint32*uint32(txIndex) + uint32(txLogIndex))
```
This function produces same results for (0, 1), (2,1) https://go.dev/play/p/z7vPLU9kBQT

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+	blockData := programEvent.BlockData
+
+	matchingFilters := lp.filters.MatchingFiltersForEncodedEvent(programEvent)
+	if matchingFilters == nil {
```
nit: it should be safe to iterate through nil seq

reductionista replied:
I thought so, but for some reason it panic's if it's nil and I try to iterate over it

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+		logs = append(logs, log)
+	}
+
+	err = lp.orm.InsertLogs(ctx, logs)
```
nit: we might want to abort insert if there is no logs to skip db transaction creation

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+
+		log.SubkeyValues = make(pq.ByteaArray, 0, len(filter.SubkeyPaths))
+		for _, path := range filter.SubkeyPaths {
+			subKeyVal, err2 := filter.DecodeSubKey(ctx, log.Data, path)
```
err2, err3 can produce diificult to find bugs, when we return err instead of err2 or 3. 
Should we also wrap them to add more context on what happened and for which log?

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+
+		log.SequenceNum = lp.filters.IncrementSeqNums(filter.ID)
+
+		expiresAt := time.Now().Add(filter.Retention).UTC()
```
Should not set expiresAt if filter.Retention is 0

dhaidashenko commented on pkg/solana/logpoller/orm.go:
```diff
@@ -200,3 +204,13 @@ func (o *DSORM) SelectLogs(ctx context.Context, start, end int64, address Public
+
+func (o *DSORM) SelectSeqNums(ctx context.Context) (map[int64]int64, error) {
+	seqNums := make(map[int64]int64)
+	query := "SELECT id, MAX(sequence_num) FROM solana.logs WHERE chain_id=%s GROUP BY id"
```
Should be filter id, otherwise we'll load whole logs table

reductionista replied:
Good catch!

dhaidashenko commented on pkg/solana/logpoller/types.go:
```diff
@@ -76,42 +82,131 @@ func (p SubkeyPaths) Value() (driver.Value, error) {
 
-const EventSignatureLength = 8
+func (v *IndexedValue) FromInt64(i int64) {
+	v.FromUint64(uint64(i + math.MaxInt64))
```
Could you add comment to explain why we need to add MaxInt64?

dhaidashenko commented on pkg/solana/logpoller/utils/anchor.go:
```diff
@@ -0,0 +1,290 @@
+	"github.com/stretchr/testify/require"
+)
+
+var ZeroAddress = [32]byte{}
```
Most of the methods in this file are not used. Do we need them?

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
 }
 
 func (lp *LogPoller) start(context.Context) error {
+	if lp.typeProvider == nil {
```
Type provider is always nil

reductionista replied:
Yeah, I replaced `lp.typeProvider` just last night with a `decoder` object attached to each filter, just forgot to remove it.

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +68,94 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+}
+
+// Process - process stream of events coming from log ingester
+func (lp *LogPoller) Process(programEvent ProgramEvent) (err error) {
```
Loader does not provide a retry mechanic for Process. If Process fails - we'll skip logs

EasterTheBunny replied:
A retry is built into the loader. If an error is returned here, the loader should retry using exponential backoff.

reductionista committed:
Add SeqNum test and fix other tests

Decoder had to be moved from Filter into a separate filters map,
otherwise it caused a lot of issues with assert.Equal for comparing
filters.

reductionista committed:
Don't set ExpireAt unless Retention is set

reductionista committed:
Move decoder creation before InsertFilter

reductionista committed:
Remove unused functions in anchor.go, add validation to TestProess test

reductionista committed:
Fix lints

reductionista committed:
Remove lp.typeProvider, generate EventIdl for test event

dhaidashenko commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +66,93 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+			TxHash:         Signature(blockData.TransactionHash),
+		}
+
+		log.Data, err = base64.StdEncoding.DecodeString(programEvent.Data)
```
errors related to decoding should be logged as critical as LP won't be able to ever process the msg. 
We might also want just to abandon this log instead of constantly retrying, as we may be able to process other msgs.

EasterTheBunny commented on pkg/solana/chain.go:
```diff
@@ -90,6 +92,7 @@ type chain struct {
 	services.StateMachine
 	id             string
 	cfg            *config.TOMLConfig
+	lp             logpoller.ILogPoller
```
Why use an interface here when we are always creating an instance of the solana log poller? A function returning an interface is very much a Go anti-pattern and should be avoided unless absolutely necessary.

In this case, we really should store and return a `*logpoller.LogPoller` instead of an interface. If we need to change it in the future, it is easy, but I think it is best to stay with convention before making exceptions.

EasterTheBunny commented on pkg/solana/chain.go:
```diff
@@ -90,6 +92,7 @@ type chain struct {
 	services.StateMachine
```
Using the following combined as service state helps clean up boilerplate:

```
	services.Service
	engine *services.Engine
```

And in the init:

```
	p.Service, p.engine = services.Config{
		Name: "LogPollerService",
		NewSubServices: func(lggr logger.Logger) []services.Service {
			p.loader = NewEncodedLogCollector(client, p, lggr)

			return []services.Service{p.loader}
		},
		Start: p.start,
		Close: p.close,
	}.NewServiceEngine(lggr)
	p.lggr = p.engine.SugaredLogger
```

reductionista replied:
Is this comment in reference to `Chain` in this file, or to `LogPoller` in the logpoller file? The example you give looks like LogPoller but you left it on the `Chain` interface.

EasterTheBunny replied:
Ahh my bad. It references `LogPoller` and may not be relevant. It looks like you did it this way so ignore my comment.

EasterTheBunny commented on pkg/solana/logpoller/log_data_parser.go:
```diff
@@ -78,8 +80,9 @@ func parseProgramLogs(logs []string) []ProgramOutput {
-					Prefix: prefixBuilder(depth),
-					Data:   dataMatches[1],
+					Program: instLogs[lastLogIdx].Program,
+					Prefix:  prefixBuilder(depth),
```
If `Prefix` isn't needed/used, we can get rid of it. The idea is that it shows overall log nesting.

reductionista replied:
Good point, as far as I know we don't need it

EasterTheBunny commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -2,40 +2,62 @@ package logpoller
+	Process(programEvent ProgramEvent) error
 }
 
 type LogPoller struct {
```
This really should be something like `Service` so that imports into other packages read as `logpoller.Service`. Using `logpoller.LogPoller` is repetative. I understand it is in an effort to be more like the `EVM` implementation, but we shouldn't copy something that is wrong for the sake of being consistently wrong.

EasterTheBunny commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -2,40 +2,62 @@ package logpoller
 }
 
-func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+func New(lggr logger.SugaredLogger, orm ORM, cl internal.Loader[client.Reader]) ILogPoller {
```
Using `New` here is great and helps provide a better coding experience. You can use `logpoller.New` instead of `logpoller.NewLogPoller` which would be repetative. However, we really should not be returning an interface here. Instead we should return a pointer type, which is Go convention.

EasterTheBunny commented on pkg/solana/logpoller/utils/anchor.go:
```diff
@@ -0,0 +1,41 @@
+package utils
```
Let's avoid a `utils` package. Maybe call it `discriminator.go` and leave it in the `logpoller` package for now?

reductionista replied:
Yeah, originally the plan was to import this file from the private repo once it's made public, so I copied it over for now. But then @dhaidashenko pointed out this is the only method actually used, so I got rid of the others. I'll just move it into the main package.

reductionista committed:
Fix decoding issues

reductionista committed:
Remove unused Prefix field from ProgramEvent

reductionista committed:
Add ExtractField function, and fix DecodeSubkeyValues

ilija42 commented on pkg/solana/logpoller/filters.go:
```diff
@@ -2,36 +2,53 @@ package logpoller
+	filtersMutex        sync.RWMutex
+	loadedFilters       atomic.Bool
+	knownPrograms       map[string]uint // fast lookup to see if a base58-encoded ProgramID matches any registered filters
+	knownDiscriminators map[string]uint // fast lookup by first 10 characters (60-bits) of a base64-encoded discriminator
```
maybe bool instead of uint

reductionista replied:
It was bool originally, had to make it a uint so it can reference count them, knowing when it's safe to delete a discriminator (once the last filter with that event sig is removed).

ilija42 commented on pkg/solana/logpoller/filters.go:
```diff
@@ -210,6 +271,45 @@ func (fl *filters) MatchingFilters(addr PublicKey, eventSignature EventSignature
+		return nil
+	}
+
+	if len(event.Data) < 12 {
```
why are we validating this, a small comment would be helpful

ilija42 commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +66,96 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+			log.SubkeyValues = append(log.SubkeyValues, indexedVal)
+		}
+
+		log.SequenceNum = lp.filters.IncrementSeqNums(filter.ID)
```
Just want to double check if this is thread safe. logs are processed in parallel, is it possible to start processing a new batch of logs with the same filterID before the previous batch is done processing

reductionista replied:
It's subtle, but yes this is thread safe.

I got confused by this too, and just tried to add a lock inside `filters.IncrementSeqNums` an hour ago due to the same worry. It resulted in a deadlock, because the mutex is already held here. The function it's locked in is the iterator returned by `fl.matchingFilters`, it gets unlocked and then locked again each time it moves to the next filter in the for loop.

I'll add some comments making this more clear, as it's not at all obvious.

reductionista committed:
MatchingFilters -> matchingFilters, and fix rest of expectedLog fields in TestProcess

ilija42 commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -44,15 +66,96 @@ func New(lggr logger.SugaredLogger, orm ORM) *LogPoller {
+		if err != nil {
+			return err
+		}
+		if len(eventData) < 8 {
```
don't we already check this above in `MatchingFiltersForEncodedEvent`

reductionista replied:
Yes, it's a redundant check which is why I made it an `AssumptionViolation` rather than an ordinary error associated with the blockchain. If the blockchain emits an event less than 8 bytes it should just move on and ignore it. But if it already matched an 8 byte discriminator and now somehow  it's less than 8 bytes, then something in our code must have gone wrong.  I added this in response to @dhaidashenko 's comment asking for more validation, which reminded me that a panic will kill the whole node instead of just reporting a critical error. It could only happen here if someone modifies the code to introduce a bug, but I figured it doesn't hurt to protect against that just in case.

ilija42 reviewed (commented):
Also made improvements on how codec is being used in Log Poller https://github.com/smartcontractkit/chainlink-solana/pull/1014

reductionista committed:
Fix lint errors, rename err2 & err3

reductionista committed:
Fix overflows, lints, use default map vals, & add early return to Process

reductionista committed:
Remove utils package, add 12-character assertion to Discriminator()

This could only fail if someone updates to a new version of
encoding/base64 which is buggy or breaks backward compatibility with the
current version. Using panic instead of returning an error, since this can
only happen during development and would need to be addressed
immediately.

reductionista committed:
Remove TODO, uncomment internal loaders

reductionista committed:
Remove ILogPoller interface, add LogPoller interface to chain.go

Also: add GetBlockWithOpts to MultiClient

reductionista committed:
Start EncodedLogCollector as soon as filters are loaded

Also:
- Double check that filters are loaded at beginning of Process()
- Add missing return from lp.loadFilters

reductionista committed:
Add comment explaining < 12 validation

reductionista committed:
Address 2 more PR comments, and remove accidentally added file

reductionista committed:
Use ch.multiClient instead of lazy-loader

reductionista committed:
revert changes to chain.go

reductionista committed:
logpoller.LogPoller -> logpoller.Service

Also: keep retrying if LogCollector fails to start

EasterTheBunny commented on pkg/solana/logpoller/discriminator.go:
```diff
@@ -0,0 +1,14 @@
+
+const DiscriminatorLength = 8
+
+func Discriminator(namespace, name string) [DiscriminatorLength]byte {
```
@ilija42 has a PR that also adds a discriminator: [here](https://github.com/smartcontractkit/chainlink-solana/pull/1014)

reductionista replied:
Yeah, that PR will go in after this one... this will use his discriminator once that gets merged.

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -310,7 +411,86 @@ func (fl *filters) LoadFilters(ctx context.Context) error {
 }
+
+func (fl *filters) DecodeSubKey(ctx context.Context, raw []byte, ID int64, subKeyPath []string) (any, error) {
+	filter, ok := fl.filtersByID[ID]
```
nit: we might want to mark this method as unsafe to signal that it should only be called from locked context of matching filters

reductionista replied:
Oh yeah, I should add that to this as well as `IncrementSeqNums`

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -210,6 +266,49 @@ func (fl *filters) MatchingFilters(addr PublicKey, eventSignature EventSignature
+// this will be called on every new event that happens on the blockchain, so it's important it returns immediately if it
+// doesn't match any registered filters.
+func (fl *filters) MatchingFiltersForEncodedEvent(event ProgramEvent) iter.Seq[Filter] {
+	if _, ok := fl.knownPrograms[event.Program]; !ok {
```
concurrent read & write to the map leads to panic

reductionista replied:
Good catch!

dhaidashenko commented on pkg/solana/logpoller/types.go:
```diff
@@ -76,42 +82,127 @@ func (p SubkeyPaths) Value() (driver.Value, error) {
-// Value implements valuer for database/sql.
-func (s EventSignature) Value() (driver.Value, error) {
-	return s[:], nil
+func NewIndexedValue(typedVal any) (iVal IndexedValue, err error) {
```
Nice work! I thought we'd have to build something super complex to support `<` `>` operators.

reductionista committed:
Make MatchFiltersForEncodedEvent thread safe

Also: add warnings to methods which are not intended to be called concurrently

reductionista committed:
Merge branch 'develop' into NONEVM-916-logpoller-process-decode

reductionista committed:
Merge branch 'develop' into NONEVM-916-logpoller-process-decode

ilija42 committed:
Merge branch 'develop' into NONEVM-916-logpoller-process-decode
